### PR TITLE
Added simple exceptions and exception handling 

### LINKED ISSUE #16: Add custom exceptions and exception handling
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
When we encounter run time errors, such as an unrecognized Primitive trying to be constructed, we need a better way to handle these errors than either ignoring them or immediately aborting the program. Simply using std::exit, std::abort, or assert does not give much information about the problem be default, and any case that includes once of those is not testable.

**Describe the solution you'd like**
We should add proper exception handling in the places in our code where a runtime error such as above could occur. If an unexpected scenario occurs, our code should throw an exception. Even if the exception is not handled by a parent function (which it likely won't be), it will cause the program to gracefully shut down and provide the user more context for what happened, and where. We can also include test cases that excessive these unexpected scenarios and can check the correct exception is thrown.

The first places that should be looked at are where we receive and cast Intents in the navigator (`rrt.cpp`), 
the static 'Primitive factory" function in the Abstract Primitive class (`primitive.h/.cpp`). There may be some other places marked with a `TODO` as well.

We should try use c++'s built-in exceptions, and choose those that are the most relevant to the error we encounter. Something like `std::invalid_argument` is probably best.

**Edit**: Make sure to update the comments for any functions that are modified to include the `@throws` label

**Describe alternatives you've considered**
LIke mentioned above, we can deal with runtime errors by either correcting for them as much as possible (eg. with default values), or aborting/exiting the program. We should not tolerate these errors as they are severe and indicated errors in the code elsewhere (ie. we should never be getting this unexpected data, where is it coming from?). We should also not abort/exit the program because without any print statements they provide no insight to why the program was stopped, and are not cases that can be tested with a unit test.

### PR DESCRIPTION
**Pull request by orayasrimokla**
### Description 
Used c++'s exception library to add exception handling. When the exception is called, it shuts down the program and provides an error message. The error message should specify where the error occurred. 

### Testing Done
Modified backend_output/main.cpp to run exceptions 

### Resolved Issues
resolves #16 

### Review Checklist
- [x] **Remove all commented out code**
- [X] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/thunderbots/software/ai/navigator/rrt/rrt.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/navigator/rrt/rrt.cpp]: Would be good to also include the name of the unrecognized Intent in the message. Makes it quicker for people to see if the error is just a typo in the name or something without having to do a full debugging session. (Can put that in place of `STAAP OR ELSE` :stuck_out_tongue_winking_eye: )

**Review by MathewMacDougall — COMMENTED** [review]
Looking great! There are still some `TODOs` in the code where exceptions need to be thrown (`ai/world/robot.cpp`, `ai/world/ball.cpp`, and others. Probably should update your branch and sweep through the code to make sure none are missed) but what's currently here looks good. I'm excited this is coming together, it looks much better than print statements :+1:

File: src/thunderbots/software/ai/primitive/primitive.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/primitive.cpp]: nit; extra space here

File: src/thunderbots/software/ai/primitive/primitive.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/primitive.cpp]: nit; extra spaces here

File: src/thunderbots/software/ai/primitive/primitive.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/primitive.cpp]: nit; extra spaces here

**Review by garethellis0 — CHANGES_REQUESTED** [review]
overall looks reasonable to me, just a few notes:
- doesn't look like we're doing any exception handling?
- could you post some examples of what a stack-trace / log message with the exceptions added here being thrown would look like?

File: src/thunderbots/software/ai/navigator/rrt/rrt.cpp
Comment (garethellis0) [reply src/thunderbots/software/ai/navigator/rrt/rrt.cpp]: As a general rule, having entertaining error messages is amusing initially, but can get a bit aggravating and unhelpful over time (`#warning` still triggers me :scream_cat:).

**Comment by garethellis0** [comment]
As a general comment, how are we going to handle things like warning vs. error'ing? Or even levels of warning? It feels like it might be helpful to have some sort of custom exception that would allow us to specify the error level, such that we could handle different error levels differently.

I'm bringing this up now because if we merge an MVP without this information, it will very rapidly get quite difficult to track down and replace all the exceptions with custom ones with the appropriate warning level.

**Comment by MathewMacDougall** [comment]
That's an interesting point, and not something I had considered.

First of all, I would prefer to avoid custom exceptions as this adds unnecessary complexity. I'd rather not nearly duplicate something that already exists. Would we have to make a custom exception for every standard exception we use? Following this paradigm, would only `warning` exceptions be caught/handled, and `error` exceptions allowed to abort the program?

I don't think exceptions need to distinguish between `warning` and `error`. IMO, exceptions should be thrown when something that is effectively an error occurs (eg. negative time values, non-existent robot IDs). AKA things that likely indicate something wrong with the code (this would include the code not accounting for certain real-world cases).

Whether we choose to catch/handle thrown exceptions or not would determine its severity. For severe cases where we really shouldn't keep running the code, we don't catch them and get a descriptive message and stack trace. This could also be accompanied by an `ERROR` log message.

Other cases can reasonably be recovered from. In these cases we should catch the exception, and can also log a `WARN` message (with the logger, #33 )

I think this would effectively accomplish what you're thinking of, without needing to define more exceptions.

**Comment by garethellis0** [comment]
I see your point here, but I don't think that quite solves the issue. The issue I see is that (as I understand it currently) errors will be caught several levels up from where the exception actually occurs. For example, if this top level exception handling catches a `std::invalid_argument`, there are cases where we can probably continue, but there are likely cases where we likely want to error and stop. 

TL;DR - how would you decide when to catch an exception or not? The same exception can indicate both terminal errors and minor warnings.

**Comment by MathewMacDougall** [comment]
So it seems what you really want is a way to differentiate exceptions of the same type (eg. differentiate `std::invalid_argument`), so that there is a clean way to catch some exceptions but not all if they are not all the same severity. This seems totally reasonable to me but it's a tradeoff between fine control and simplicity. I'd be interested in how we might achieve this. I don't like the idea of making our own versions of some/most of the std exceptions.

Off the top of my head I don't have any examples of situations where a function might throw some recoverable and some unrecoverable exceptions, but obviously that's not future-proof. However I do think that exceptions will/should be caught very close to where they are thrown. Using an invalid argument as an example, that would be something that should be detected and handled right away by the calling code if it's recoverable. It's harder to do that kind of recovery from several levels up.

This doesn't seem like an unreasonable thing to want to do, so I'm a bit surprised I didn't find any libraries (but I didn't look past the first page of Google). Perhaps there are other ways to differentiate exceptions of the same type (that isn't a string comparison of the exception message).

**Comment by garethellis0** [comment]
Perhaps the exception handling at the top level should just be to stop, because if there is an exception that this not handled before then we don't really know of any reasonable recovery action?

**Comment by MathewMacDougall** [comment]
I agree with that, but that doesn't solve the finer control problem of different exception severities.

**Comment by MathewMacDougall** [comment]
I no longer think I fully understand what you want out of this extra exception severity. Finer control over exception severity seems mostly unrelated to how the AI lives/dies by exceptions at the highest level. My understanding is that we simply may want some exceptions to be caught, and others to propagate to terminate the system, even if these exceptions are of the same type (eg. `std::invalid_argument`)

in general I agree with the paradigm of "ai should never kill itself by choice". I think achieving this was unconsciously part of the goal of moving to use exceptions in the first place, since they allow us to catch and handle things that traditionally would take down the AI.

I think we will be fine just using the standard exceptions for 2 reasons:
1. I realistically don't see modules of our code throwing exceptions of the same type where some are recoverable and some are not (which is the case we would want that extra fine control to distinguish them). This leads to my second point...
2. I believe that when used properly, the standard exceptions inherently contain some measure of severity. The nature of the exception dictates whether or not it is recoverable. For example, `std::invalid_argument` might be thrown if a function is given a value it doesn't expect, like a negative value. This is easy to recover from _if we want_. IMO this generalizes to effectively all cases where a function receives an invalid argument. The range of valid arguments is known so it's reasonable to recover from this if desired.

    On the other hand, an exception such as `file not found` is inherently not something recoverable (assuming we're opening a file for reading). There's not really a good course of action to correct this case, so the system should probably abort.

    If a function/module ever finds itself in the case where it's throwing so many exceptions that we truly would want to recover from some invalid arguments but not others, the function is too big and not performing a single unit of work. In this case we should refactor the function and handle the exceptions of each new function individually, rather than doing very detailed exception handling on a single large function.

    (end mini-rant)

Looping in @EvanMorcom @akhilveeraghanta @jared-paul @jessicahe21 @jordanzliu for some other opinions. We should try settle on something soon so we can move forwards.

**Comment by garethellis0** [comment]
Sorry, I was somewhat talking about two different things (measure for exception severity and how to handle exceptions that make it to top level), but I believe we're on the same page. 

At this point my thoughts are this:
- use standard exceptions
- if an exception in the ai loop propagates to the top level, log a warning and skip _the current iteration_ of the loop (we could also abort for specific exceptions if we wanted to)

This means that if we have recovery actions for the exceptions in the code, then great, it doesn't make it to top level. If we don't, then it goes to top level, we get a warning, and we just go to the next iteration of the loop.

**Comment by MathewMacDougall** [comment]
Sounds good to me. We'll just want to log those top-level warning _very_ *loudly* so it's clear we've recovered/skipped something potentially severe.

**Comment by MathewMacDougall** [comment]
@orayasrimokla you should be able to continue your changes using the standard exceptions like you were doing. And based on the discussions above, handle most exceptions as close to where they are thrown if possible, otherwise catch them at the main `ai` loop and skip that iteration.

**Comment by orayasrimokla** [comment]
Will be updated soon!

**Comment by MathewMacDougall** [comment]
@orayasrimokla any updates for this?

**Comment by orayasrimokla** [comment]
sorry not yet :( I've been taking quite a bit of work home.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Existing changes look good. There are still some `#TODOs` that need exceptions in there place:
* Several "negative time" exceptions in `world/robot.cpp`
* An exception for multiple robots with the same id in `world/team.cpp`
An easy way to see the `#TODOs` in the code is to click on `View -> Tool Windows -> TODO` from the CLion menu bar

In addition, functions that now throw an exception should have a `@throws` entry added to their javadoc comment, giving a brief description of the exception thrown and it what cases it's thrown. Eg. for the Ball's estimation functions "@throws std::invalid_argument if the given timestamp is older than the timestamp of when the ball was last updated (a time in the past was given)"

Finally, there are a few unit tests that should be updated now that we are throwing exceptions. The great thing about this, is that we can use them in unit tests to make sure they are being thrown when we expect. For example, in the unit tests for `ball.cpp` there is an empty test called `update_state_with_past_timestamp` that needs to be implemented. It should try update the ball state with a timestamp from the past, and that a `std::invalid_argument` is thrown. This can be done with gtest's `ASSERT_THROW` statement. More info can be found [here](https://github.com/abseil/googletest/blob/master/googletest/docs/advanced.md#exception-assertions)

**Comment by MathewMacDougall** [comment]
@orayasrimokla any updates?

**Comment by orayasrimokla** [comment]
@MathewMacDougall Working on unit testing today!

**Comment by MathewMacDougall** [comment]
Overall looks good. @orayasrimokla your branch will need to be updated before it can be merged, and that should trigger a new CI build since it looks like the existing one failed due to failing to clone (likely a network issue when it was trying to run).

@garethellis0 for re-review

File: src/thunderbots/software/ai/main.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp]: Hmm, should we be catching specific exceptions here? Or all exceptions? I had thought the idea was to make sure that no matter what, we could carry on if a given iteration of AI threw an exception. @MathewMacDougall ?

File: src/thunderbots/software/ai/world/team.h
Comment (garethellis0) [inline src/thunderbots/software/ai/world/team.h]:
``` suggestion
     * @throws std::invalid_argument if the goalie is assigned to a robot that is not on the team
```

File: src/thunderbots/software/test/world/robot.cpp
Comment (garethellis0) [inline src/thunderbots/software/test/world/robot.cpp:278]: Nice job sticking these tests in!

**Review by garethellis0 — COMMENTED** [review]
I'd like to also note that I'm thrilled to see a lot of little hacks get resolved with this PR!

File: src/thunderbots/software/ai/main.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/main.cpp]: No, the AI shouldn't be 100% fault-tolerant. There still may be cases where something is so terribly wrong we want to abort.

Catching specific exceptions is a way of implementing the "severity" we talked about above, where we can allow some exceptions to be caught and recovered from and other to terminate the AI (or be caught at a higher level).

Comment (garethellis0) [reply src/thunderbots/software/ai/main.cpp]: Fair enough, I agree with this!

**Comment by garethellis0** [comment]
@orayasrimokla - good job dealing with those comments! Just as a note for the future, could you please leave a quick reply to each comment (even if it's just as simple as "changed", or "not changed, because ...") and leaving them unresolved? This helps me go through comments later, particularly in large PR's with several rounds of review.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
Uh oh. Looks like something weird happened when the branch was updated. There's commits that are already in master showing up here as new commits. @orayasrimokla what steps did you take to update your branch?

File: src/thunderbots/software/CMakeLists.txt
Comment (garethellis0) [inline src/thunderbots/software/CMakeLists.txt]: This basically looks good to me, except for the merge conflicts in this file. Once we resolve those, should be good to go! Let me know if you want any help.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:. Great work, and Merry Christmas :christmas_tree: :tada:

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Whoops I didn't notice that CI didn't pass. Looks like there's a unit test that isn't catching/expecting an exception to be thrown (something with the goalie id), and formatting will need to be updated afterwards.

Otherwise it all looks good :+1:

**Review by garethellis0 — DISMISSED** [review]
LGTM (once CI passes anyhow) :+1: :tada: Merry Christmas! :christmas_tree:

**Comment by MathewMacDougall** [comment]
@orayasrimokla how's progress? ETA on a branch update with the fixes we found on Saturday? That should fix the test errors. Your branch will need to be updated to fix conflicts as well.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
@orayasrimokla changes look good :+1: . Just looks like you'll need to update your branch from `upstream master`, and run the formatting script.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1: Great work!

**Review by henryymliu — APPROVED** [review]

**Comment by orayasrimokla** [comment]
AyOOOoooOOOo

**Pull request merged**
